### PR TITLE
fix(providers/amazon): handle missing LogUri in emr describe_cluster API response

### LINKED ISSUE #31480: Missing LogUri from emr describe-cluster API when executing EmrCreateJobFlowOperator
### Apache Airflow version

main (development)

### What happened

Encounter the following error when executing `EmrCreateJobFlowOperator`

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/airflow/providers/amazon/aws/operators/emr.py", line 695, in execute
    log_uri=get_log_uri(emr_client=self._emr_hook.conn, job_flow_id=self._job_flow_id),
  File "/usr/local/lib/python3.10/site-packages/airflow/providers/amazon/aws/links/emr.py", line 61, in get_log_uri
    log_uri = S3Hook.parse_s3_url(response["Cluster"]["LogUri"])
KeyError: 'LogUri'
```

According to the [this document](https://docs.aws.amazon.com/cli/latest/reference/emr/describe-cluster.html), it seems we might not always be able to get `["Cluster"]["LogUri"]`  and we encounter errors after the release in https://github.com/apache/airflow/issues/31322.

### What you think should happen instead

The `EmrCreateJobFlowOperator` should finish execution without error.

### How to reproduce

1. `git clone github.com/apache/airflow/`
2. `cd airflow`
3. `git checkout c082aec089405ed0399cfee548011b0520be0011` (the main branch when I found this issue)
4. Add the following DAG to `files/dags/` and name it as `example_emr.py`

```python
import os
from datetime import datetime, timedelta

from airflow import DAG
from airflow.providers.amazon.aws.operators.emr import EmrCreateJobFlowOperator, EmrTerminateJobFlowOperator


JOB_FLOW_OVERRIDES = {
    "Name": "example_emr_sensor_cluster",
    "ReleaseLabel": "emr-5.29.0",
    "Applications": [{"Name": "Spark"}],
    "Instances": {
        "InstanceGroups": [
            {
                "Name": "Primary node",
                "Market": "ON_DEMAND",
                "InstanceRole": "MASTER",
                "InstanceType": "m4.large",
                "InstanceCount": 1,
            },
        ],
        "KeepJobFlowAliveWhenNoSteps": False,
        "TerminationProtected": False,
    },
    "JobFlowRole": "EMR_EC2_DefaultRole"
    "ServiceRole": "EMR_DefaultRole",
}

DEFAULT_ARGS = {
    "execution_timeout": timedelta(hours=6),
    "retries": 2,
    "retry_delay": 60,
}


with DAG(
    dag_id="example_emr_sensor",
    schedule=None,
    start_date=datetime(2022, 1, 1),
    default_args=DEFAULT_ARGS,
    catchup=False,
) as dag:
    create_job_flow = EmrCreateJobFlowOperator(
        task_id="create_job_flow",
        job_flow_overrides=JOB_FLOW_OVERRIDES,
        aws_conn_id="aws_default",
    )

    remove_job_flow = EmrTerminateJobFlowOperator(
        task_id="remove_job_flow",
        job_flow_id=create_job_flow.output,
        aws_conn_id="aws_default",
        trigger_rule="all_done",
    )

    create_job_flow >> remove_job_flow
```

5. `breeze --python 3.8 --backend sqlite start-airflow`
6. Trigger the DAG from web UI

### Operating System

mac OS 13.4

### Versions of Apache Airflow Providers

_No response_

### Deployment

Astronomer

### Deployment details

_No response_

### Anything else

https://github.com/apache/airflow/issues/31322#issuecomment-1556876252

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by Lee-W**
According to the [this document](https://docs.aws.amazon.com/cli/latest/reference/emr/describe-cluster.html), it seems we might not always be able to get ["Cluster"]["LogUri"], and we encounter errors after the release in https://github.com/apache/airflow/issues/31322. In this PR, I set the return value of `get_log_uri` to `None` if we cannot get `LogUri`

Closes: #31480

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by potiuk — APPROVED** [review]
LGTM

**Comment by ferruzzi** [comment]
If this returns none, what happens with the displayed link that gets generated from this value?

**Comment by Lee-W** [comment]
@ferruzzi Hi, I just tested it. It will link to a None bucket.

**Comment by Lee-W** [comment]
Should I disable it if LogUri not found?

**Comment by Lee-W** [comment]
Hi @potiuk , I encounter the following CI failure. But it doesn't seem to be something I introduce in this PR. Should I fix it in this PR? If so could you kindly point out where this is from? Thanks!

```
Lint OpenAPI using openapi-spec-validator..............................................Failed
- hook id: lint-openapi
- exit code: 1

'_io.BufferedReader' object has no attribute 'decode'
```

**Comment by pankajkoti** [comment]
> Hi @potiuk , I encounter the following CI failure. But it doesn't seem to be something I introduce in this PR. Should I fix it in this PR? If so, could you kindly point out where this is from? Thanks!
> 
> ```
> Lint OpenAPI using openapi-spec-validator..............................................Failed
> - hook id: lint-openapi
> - exit code: 1
> 
> '_io.BufferedReader' object has no attribute 'decode'
> ```

I see a similar issue in my PR and have added a comment regarding this https://github.com/apache/airflow/pull/31201#issuecomment-1559977621

Apparently, @uranusjr did try to reproduce this with the PR https://github.com/apache/airflow/pull/31489 as this was observed in `main` too, but we failed to reproduce it there and `main` also seemed to run fine later. This is not reproducible locally and looks like some weird issue. I would say we could wait for a while.

File: airflow/providers/amazon/aws/links/emr.py
Comment (hussein-awala) [inline airflow/providers/amazon/aws/links/emr.py]: When cluster is provided, we get it by the same method but we provide it to avoid querying the API twice, so we can have the same exception.
We can simplify this method by using this code:
```suggestion
    cluster_info = (cluster or emr_client.describe_cluster(ClusterId=job_flow_id))["Cluster"]
    if "LogUri" not in cluster_info:
        return
    log_uri = S3Hook.parse_s3_url(cluster_info["LogUri"])
    return "/".join(log_uri)
```

**Review by hussein-awala — CHANGES_REQUESTED** [review]

File: airflow/providers/amazon/aws/links/emr.py
Comment (Lee-W) [reply airflow/providers/amazon/aws/links/emr.py]: Thanks for your suggestion! I just updated it

**Review by hussein-awala — APPROVED** [review]
LGTM

**Comment by hussein-awala** [comment]
> @ferruzzi Hi, I just tested it. It will link to a None bucket.
Should I disable it if LogUri not found?

IMO this is not a requirement to merge this fix, but it's a nice enhancement for the AWS operators links which can be implemented separately.
@ferruzzi @o-nikolas wdyt?

**Comment by ferruzzi** [comment]
> IMO this is not a requirement to merge this fix, but it's a nice enhancement for the AWS operators links which can be implemented separately.
@ferruzzi @o-nikolas wdyt?

I'm not thrilled with the idea of knowingly vending a broken link.  What about a compromise for now where if the link isn't available, return a string along the lines of "N/A"?I Love the idea of disabling the link if it's not available, but a static string message would be better than a dead link IMHO

**Comment by pankajkoti** [comment]
> > IMO this is not a requirement to merge this fix, but it's a nice enhancement for the AWS operators links which can be implemented separately.
> > @ferruzzi @o-nikolas wdyt?
> 
> I'm not thrilled with the idea of knowingly vending a broken link. What about a compromise for now where if the link isn't available, return a string along the lines of "N/A"?I Love the idea of disabling the link if it's not available, but a static string message would be better than a dead link IMHO

+1 to this idea and perhaps some more description in the string to include something like "that you have disabled logging and hence ...." based upon the comment https://github.com/apache/airflow/issues/31480#issuecomment-1560723485 might hint the user why it is N/A, no?

**Comment by Lee-W** [comment]
Hi @ferruzzi @pankajkoti , thanks for your suggestions! I'm not sure how I can add those descriptions but I disabled it if the link is not avaiable.

image

**Comment by ferruzzi** [comment]
> Hi @ferruzzi @pankajkoti , thanks for your suggestions! I'm not sure how I can add those descriptions but I disabled it if the link is not available.

I love it, thanks for doing that.   I _think_ pankajkoti's suggestion was around the phrasing where you put "No URL found for EMR Cluster logs".  My initial suggestion was super vague and I _think_ they were suggesting it be more descriptive.  I think you have it covered, but I'll leave that to them.

**Review by ferruzzi — APPROVED** [review]
I'll approve pending passing CI.  I have seem that test fail on occasion for a timeout, try bumping it again.

**Review by pankajkoti — COMMENTED** [review]
Should we add a test here for this? At least one with covering that if cluster_info does not contain LogUri, it does not raise an exception anymore?

Looks good to me otherwise.

**Comment by jedcunningham** [comment]
LGTM, but missing test coverage.

**Comment by Lee-W** [comment]
@pankajkoti Thanks for your suggestion! I just added it.

**Review by pankajkoti — APPROVED** [review]

**Pull request merged**
